### ISSUE TITLE
Segfault when currying function without really curring (?)

### ISSUE THREAD

**Issue by kirbyfan64** [kevinlawler/kona#356]
Kona:

```
ryan@DevPC-LX:~/plang$ k
K Console - Enter \ for help

  {x}[;] 
Segmentation fault (core dumped)
ryan@DevPC-LX:~/plang$
```

K2:

```
ryan@DevPC-LX:~/plang$ k2
K 2.8 2000-10-10 Copyright (C) 1993-2000 Kx Systems
Evaluation. Not for commercial use. 
\ for help. \\ to exit.

  {x}[;]
{x}[;]
  
```

**Comment by tavmem** [comment]
In k2.8
```
  {x}[;]
{x}[;]
  {x}[1;]
{x}[1;]
  {x}[;1]
{x}[;1]
```
In Kona
```
  {x}[;]
Segmentation fault
  {x}[1;]
valence error
  {x}[;1]
Segmentation fault
```
What do we want Kona to yield in these 3 cases?
Seems to me that "valence error" may be appropriate (in all 3 cases).
Are any of these forms useful?

in k2.8:
```
  f:{x}[;]
  f 1
valence error
f 1
^
>  \
  f[1;2]
valence error
f[1;2]
^
> 
```

**Comment by kirbyfan64** [comment]
@tavmem Well, regardless of which way is correct, I doubt a segfault could ever be ideal behavior. :)

**Comment by tavmem** [comment]
No, I am not suggesting segfault as a viable option.

Option 1: (like k2.8) allow each form to parse as valid.  Then report "valence error" when you try to use it.

Option 2:  Report a "valence error" as a result of the parse.

My question relates to why k2.8 reports valence error at "run time" instead of at "parse time".
Is there some utility in allowing a valid parse?  Am I missing something?  Can the form be used in some context with valid results?

**Comment by kirbyfan64** [comment]
@tavmem That was a joke. :O

I think the difference is that k2 doesn't check the argument counts when creating a function partial (or whatever it's called). It simply binds the arguments. Kona does check the argument counts. I would prefer for it to be a runtime error due to logic reasons, but give an error during the actual execution of the partialing/currying/whatever. I don't know why it is k2 would wait till call time...

**Comment by JohnEarnest** [comment]
I don't think it is possible, in general, to validate currying lists at parse time, because a function may be referred to indirectly through a variable or as the result of an expression- the valence of the function could thus only be known at runtime:

      f: {x+y}
    {x+y}
      f[2;]
    {x+y}[2;]
      f: {x}
    {x}
      f[2;]
    valence error

In the specific case of a function literal followed by a curry list you _can_ check that they match at parse time, but it is inconsistent with currying in other contexts.

**Comment by bakul** [comment]
On Sun, 21 Jun 2015 14:11:22 PDT Ryan Gonzalez <notifications@github.com> wrote:
> 
> I don't know why it is k2 would wait till call time...

Consider:

      f:{x+y}[1;2]
      f
    2

      f:{x+y}[1;2;3]
    valenc error
    f:{x+y}[1;2;3]
      ^

      f:{x+y}[1;2;]
      f
    {x+y}[1;2;]

      f:{x+y}[1;]
      f
    {x+y}[1;]
      f 2
    3

      f:{x+y}[1;]
      f
    {x+y}[;2]
      f 3
    5

      f:{x+y+z}[1;;]
      f
    {x+y+z}[1;;]
      g:f[2]
      g
    {x+y+z}[1;;][2]
      g 3
    6

What these last few example reveal is that if an arg is
omitted, actual arg evaluation is left after sufficient
additional args have been supplied and inserted in the correct
place. Only then is the function actually evaluated. And this
is the point where it is discovered args do not match up.

In other words {...} is cracked open only once, after all the
args are available and only at that point can you discover
# of args don't match.

**Comment by tavmem** [comment]
Yes, that makes a lot of sense.
As you pointed out yesterday, some of what makes k2 and k3 faster than Kona is that Arthur has identified places where parsing can be eliminated, whereas Kona sometimes parses more than it needs to (parsing the same code multiple times).

The better option (which should lead to a faster product) is to go the k2/k3 route.  However, that would probably require big changes to the Kona parsing strategy.

In the short tem, to eliminate the segfault condition, I will probably just raise a "valence error" during the Kona parse, and open an addtional issue to revisit the Kona parsing strategy.
